### PR TITLE
feat!: make on create an async function

### PR DESCRIPTION
**Pull request by bartoszherba**
## Description
The goal of this task was to make the `onCreate` an async function. To achieve that we must introduce BC change. The most intrusive change is that the`apiClientFactory` also becomes an async function.

Because the `onCreate` function is strongly coupled with the factory and lifecycles, there is no other way to make it an async function but to make the whole creation process asynchronous.

As a migration step, all `onCreate` functions in `index.server.ts` in all integrations must be updated to the async function. This is a huge change and will have an impact on all integrations.

## How Has This Been Tested?
I tested that for the Magento integration. Once the `onCreate` function was updated to the async function build and app started to work properly as before without any noticeable bugs or problems.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.

#### Changelog
- [x] I have updated the Changelog ([V1](https://github.com/DivanteLtd/vue-storefront/blob/develop/CHANGELOG.md)) [v2](https://docs-next.vuestorefront.io/contributing/creating-changelog.html) and mentioned all breaking changes in the public API.
- [x] I have documented all new public APIs and made changes to existing docs mentioning the parts I've changed so they're up to date.

#### Tests
- [ ] I have written test cases for my code
- [x] I have tested my Pull Request on production build and (to my knowledge) it works without any issues
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

> I tested manually my code, and it works well with both:
- [x] Default Theme
- [ ] Capybara Theme

#### Code standards
- [ ] My code follows the code style of this project.

- [ ] I have followed [naming conventions](https://github.com/kettanaito/naming-cheatsheet)

#### Docs
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: packages/core/src/factories/apiClientFactory/index.ts [hunk 1]
```diff
@@ -23,7 +23,7 @@ const apiClientFactory = <ALL_SETTINGS extends ApiClientConfig, ALL_FUNCTIONS>(f
       .filter(ext => isFn(ext.beforeCreate))
       .reduce((prev, curr) => curr.beforeCreate({ configuration: prev }), config);
 
-    const settings = factoryParams.onCreate ? factoryParams.onCreate(_config) : { config, client: config.client };
+    const settings = factoryParams.onCreate ? await factoryParams.onCreate(_config) : { config, client: config.client };
```
Comment (tlos-vsf) [inline packages/core/src/factories/apiClientFactory/index.ts:26]: Here is the change I suggested during our daily, exposing the `req` and `res` properties on `onCreate`:

```suggestion
    const settings = factoryParams.onCreate ? await factoryParams.onCreate(_config, this.middleware) : { config, client: config.client };
```

In general, according to the `ApiClientFactoryParams` type, the `onCreate` function should take 2 params:

```
export interface ApiClientFactoryParams<T, F = any> {
  ...
  onCreate: (config: T, headers?: Record<string, string>) => Promise<{ config: T; client: any }>;
  ...
}
```

Without the above change, `headers` are `undefined` and for the purpose of multistore functionality, we need the `req` to pick up the origin hostname. In the **dev** environment, we also need `res`, as there are problems with CORS, and we need a proper `Access-Control-Allow-Origin` header dynamically generated and appended to the response based on the hostname where the request came from.

**Review by tlos-vsf — CHANGES_REQUESTED** [review]
Great work! Your implementation aligns with my PoC for multistore functionality. There is only one thing missing, which I suggested in the comment below.

File: packages/core/src/factories/apiClientFactory/index.ts (hunk 1 above)
Comment (bartoszherba) [reply packages/core/src/factories/apiClientFactory/index.ts:26]: From what I understand, we need an optional req and/or res object. 
Can't we use already existing hooks like `beforeCreate` or `hook` function to set/get whatever you need?
The fact that we do not pass headers there is one thing, but passing the whole middleware will make everything even vaguer. I really do not want to do that. Middleware is a singleton for that process and so are the req and res objects on the middleware, both available in hooks. Please, consider such a solution.

**Review by tlos-vsf — APPROVED** [review]
Looks good to me. I agree that retrieving `req` and `res` objects through API hooks (namely `beforeCreate`) is a cleaner solution, so I'm dropping my initial suggestion regarding exposing the middleware object to the `onCreate`. Thank you for your work on that issue!

**Review by Frodigo — APPROVED** [review]

File: packages/docs/integrate/integration-guide.md [hunk 2]
```diff
@@ -131,29 +131,32 @@ yarn add axios
...
 
-  return {
-    config: settings,
-    client: settings.client
-  };
+  const config = buildConfig(settings);
+
+  return new Promise((resolve) => {
+    resolve({ config, client: settings.client });
+  });
```
Comment (Fifciu) [inline packages/docs/integrate/integration-guide.md:159]:
```suggestion
  return Promise.resolve({ config, client: settings.client });
```

File: packages/core/src/types.ts [hunk 3]
```diff
@@ -852,7 +852,7 @@ export interface MiddlewareConfig {
 export interface ApiClientFactoryParams<T, F = any> {
   api: F;
   isProxy?: boolean;
-  onCreate: (config: T, headers?: Record<string, string>) => { config: T; client: any };
+  onCreate: (config: T, headers?: Record<string, string>) => Promise<{ config: T; client: any }>;
```
Comment (Fifciu) [inline packages/core/src/types.ts:855]: I know it's not a scope of this task. But maybe we should remove `headers` argument as we already changing smth here? I do not see any place where it's used. Do I miss something?

Comment (bartoszherba) [reply packages/core/src/types.ts:855]: No, you are not missing anything, it looks like an old code or a dead code. I will remove it then.

File: packages/docs/integrate/integration-guide.md (hunk 2 above)
Comment (bartoszherba) [reply packages/docs/integrate/integration-guide.md:159]: ok, updated

**Review by Fifciu — APPROVED** [review]

**Review by tlos-vsf — APPROVED** [review]

**Review by WojtekTheWebDev — CHANGES_REQUESTED** [review]
Please give me a day more to review ;)

File: packages/core/src/types.ts [hunk 4]
```diff
@@ -852,7 +852,7 @@ export interface MiddlewareConfig {
 export interface ApiClientFactoryParams<T, F = any> {
   api: F;
   isProxy?: boolean;
-  onCreate: (config: T, headers?: Record<string, string>) => { config: T; client: any };
+  onCreate: (config: T) => Promise<{ config: T; client: any }>;
```
Comment (WojtekTheWebDev) [inline packages/core/src/types.ts:855]: Why are the `headers` removed here?

Comment (Fifciu) [reply packages/core/src/types.ts:855]: @WojtekTheWebDev look here: https://github.com/vuestorefront/vue-storefront/pull/6807#discussion_r977422605

Comment (bartoszherba) [reply packages/core/src/types.ts:855]: Can we resolve this?

Comment (WojtekTheWebDev) [reply packages/core/src/types.ts:855]: Thx for explaining

**Review by WojtekTheWebDev — APPROVED** [review]
LGTM, now we need to prepare the release

**Comment by bartoszherba** [comment]
@WojtekTheWebDev I can prepare the release, but do we have a procedure of how to do that?

**Comment by WojtekTheWebDev** [comment]
@bartoszherba I've changed base branch to chore/release-2.8.0-canary so we can merge this one. I'll take care of final release

**Pull request merged**